src/sql/src/query_model/rewrite_engine.rs

```diff
@@ -965,3 +966,33 @@ impl Rule for WindowingToSelect {
         b.box_type = BoxType::Select(Select::new());
     }
 }
+
+/// A Grouping box without aggregations is a Select box with
+/// [DistinctOperation::Enforce]
+#[derive(Default)]
+struct GroupingToDistinct;
+
+impl Rule for GroupingToDistinct {
+    fn name(&self) -> &'static str {
+        "GroupingToDistinct"
+    }
+
+    fn rule_type(&self) -> RuleType {
+        RuleType::PostOrder
+    }
+
+    fn condition(&mut self, model: &Model, box_id: BoxId) -> bool {
+        let b = model.get_box(box_id);
+        if let BoxType::Grouping(g) = &b.box_type {
+            b.columns.iter().all(|c| g.key.contains(&c.expr))
```

Comment:
This could just be:

```
b.columns.iter().all(|c| !matches!(c.expr, BoxScalarExpr::Aggregate { .. }))
```

because elements in the projection of a grouping box are either column references from the input quantifier that are also part of the grouping key or aggregate expressions.

Reply:
The action is correct for the `condition` you have implemented. However, we should convert any grouping box without aggregates into a select box. And, a grouping box may not project all the columns  from the grouping key, we must convert it into a chain of two Select boxes: one with `Distinct::Enforce` projecting all the columns in the grouping key and another one on top of that one to match the projection of the grouping box.

The action then would be:

1. create a new select box with distinct enforce
2. move the quantifier of the grouping box to that box
3. add all the columns in the grouping key to the projection of the select box
4. convert the grouping box into a select box (without distinct enforce)
5. add a foreach quantifier to the parent select box ranging over the box created in 1.
6. add the columns of the quantifier created above in the order they appeared in the grouping box into the projection of the converted box

The action for the condition method I've pasted above would be:

```
let grouping_key = if let BoxType::Grouping(grouping) = &model.get_box(box_id).box_type {
    grouping.key.clone()
} else {
    panic!();
};
// 1
let distinct_select_id = model.make_select_box();
// 2
model.swap_quantifiers(distinct_select_id, box_id);
// 3
for c in grouping_key.iter() {
    model.get_mut_box(distinct_select_id).add_column(c.clone());
}

// 5
let new_quantifier_id = model.make_quantifier(QuantifierType::Foreach, distinct_select_id, box_id);

// 4
let mut b = model.get_mut_box(box_id);
b.box_type = BoxType::Select(Select::new());

// 6
for c in b.columns.iter_mut() {
    c.expr = BoxScalarExpr::ColumnReference(ColumnReference {
        quantifier_id : new_quantifier_id,
        position: groping_key.iter().position(|k| c.expr == k).unwrap()
    });
}
```

Again, your implementation is correct, but it leaves behind some cases where the grouping box can be replaced by a chain of two Select boxes that could potentially be squashed by SelectMerge transform, if the conditions are met.

Reply:
BTW there is an example of this in the handover doc.

Reply:
I tried out writing my own version of the 6 steps of the action, mostly to see why all those six steps need to exist and why you chose to convert the current box into the outer select instead of the inner select.